src/connection.ts

```diff
@@ -1774,6 +1777,442 @@ class Connection extends EventEmitter {
+      this.addToMessageBuffer(data);
+    }
+
+    const preloginPayload = new PreloginPayload(this.messageBuffer);
+    this.debug.payload(function() {
+      return preloginPayload.toString('  ');
+    });
+
+    if (preloginPayload.fedAuthRequired === 1) {
+      this.fedAuthRequired = true;
+    }
+
+    if (preloginPayload.encryptionString === 'ON' || preloginPayload.encryptionString === 'REQ') {
+      if (!this.config.options.encrypt) {
+        throw new ConnectionError("Server requires encryption, set 'encrypt' config option to true.", 'EENCRYPT');
+      }
+
+      this.messageIo.startTls(this.secureContext, this.routingData?.server ?? this.config.server, this.config.options.trustServerCertificate);
+      this.transitionTo(this.STATE.SENT_TLSSSLNEGOTIATION);
+
+      while (!this.messageIo.tlsNegotiationComplete) {
+        let message;
+        try {
+          message = await this.messageIo.readMessage(signal);
+        } catch (err: any) {
+          if (err instanceof Error && err.name === 'AbortError') {
+            throw err;
+          }
+
+          const message = `Failed to connect to ${this.config.server}:${this.config.options.port} - ${err.message}`;
+          this.debug.log(message);
+          throw new ConnectionError(message, 'ESOCKET');
+        }
+
+        addAbortSignal(signal, message);
+        for await (const data of message) {
+          this.messageIo.tlsHandshakeData(data);
+        }
+      }
+    }
+
+    this.sendLogin7Packet();
+
+    const { authentication } = this.config;
+    if (authentication.type === 'azure-active-directory-password' || authentication.type === 'azure-active-directory-msi-vm' || authentication.type === 'azure-active-directory-msi-app-service' || authentication.type === 'azure-active-directory-service-principal-secret' || authentication.type === 'azure-active-directory-default') {
+      this.transitionTo(this.STATE.SENT_LOGIN7_WITH_FEDAUTH);
+      return this.handleLogin7WithFedauthResponse(signal);
+    } else if (authentication.type === 'ntlm') {
+      this.transitionTo(this.STATE.SENT_LOGIN7_WITH_NTLM);
+      return this.handleLogin7WithNTLMResponse(signal);
+    } else {
+      this.transitionTo(this.STATE.SENT_LOGIN7_WITH_STANDARD_LOGIN);
+      return this.handleLogin7WithStandardLoginResponse(signal);
+    }
+  }
+
+  async handleLogin7WithNTLMResponse(signal: AbortSignal) {
+    while (true) {
+      let message;
+      try {
+        message = await this.messageIo.readMessage(signal);
+      } catch (err: any) {
+        if (err instanceof Error && err.name === 'AbortError') {
+          throw err;
+        }
+
+        const message = `Connection lost - ${err.message}`;
+        this.debug.log(message);
+        throw new ConnectionError(message, 'ESOCKET');
+      }
+
+      const handler = new Login7TokenHandler(this);
+      const tokenStreamParser = this.createTokenStreamParser(message, handler);
+
+      await once(tokenStreamParser, 'end');
```

Comment:
The `tokenStreamParser` does not honor the `AbortSignal` yet. Potentially `addAbortSignal(message, signal)` before the `this.createTokenStreamParser` might be enough, but I haven't tried that yet.